### PR TITLE
Better error message for no GPU or incompatible GPU

### PR DESCRIPTION
**Pull request by zkhatami**
To give user some clue what's happening if the program gets compiled on a node with no GPU or if it gets compiled with different compute capability than the one it's running on. In both scenarios no good error message were produced before. The proposed changes will improve the user experience and make it easier for users to troubleshoot problems.

This fix is for addressing the issue#1785 reported on Thrust https://github.com/NVIDIA/thrust/issues/1785

### CODE REVIEW

**Comment by zkhatami** [comment]
From issue#1785 on thrust (https://github.com/NVIDIA/thrust/issues/1785), for this small test case:

`#include <thrust/device_vector.h>
#include <thrust/sort.h>
int main() {
  thrust::device_vector<int> dv;
  thrust::sort(dv.begin(), dv.end());
}`

when compiled with -gpu=cc60 and then run it on a system with cc80, the error message would be:
`
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  radix_sort: failed on 1st step: cudaErrorUnsupportedPtxVersion: the provided PTX was compiled with an unsupported toolchain.
Aborted
`

This doesn't help user to understand what's happening. I tried to address it in this change so that better message will show up:
`
Incompatible GPU: you are trying to run this program on sm_80, different from the one that it was compiled for
`

File: cub/util_device.cuh [hunk 1]
```diff
@@ -424,6 +455,9 @@ CUB_RUNTIME_FUNCTION inline cudaError_t PtxVersion(int &ptx_version)
     NV_IS_HOST,
     (
       auto const device  = CurrentDevice();
+      if (device < 0) {
+        printf("No GPU is available\n");
+      }
```
Comment (jrhemstad) [inline cub/util_device.cuh:460]: Was this for debugging? I don't think we want to keep this.

File: cub/util_device.cuh [hunk 2]
```diff
@@ -348,6 +366,19 @@ CUB_RUNTIME_FUNCTION inline cudaError_t PtxVersionUncached(int& ptx_version)
...
+        // Any failure means either no GPU is available or the provided device
+        // binary does not match the generated function code
+        if (result != cudaSuccess) {
+            int sm_version;
+            int device;
+            if (!CubDebug(cudaGetDevice(&device))) {
+                if (!CubDebug(SmVersionUncached(sm_version, device))) {
+                    printf("Incompatible GPU: you are trying to run this program on sm_%d%d, "
+                            "different from the one that it was compiled for\n",
+                            sm_version/100, (sm_version%100)/10);
```
Comment (jrhemstad) [inline cub/util_device.cuh:378]: `printf` isn't a very robust or canonical way of reporting errors. This should really be an exception, but CUB doesn't currently throw exceptions.

**Review by jrhemstad — COMMENTED** [review]
The other problem with this approach is it is dependent on our current strategy of using `cudaFuncGetAttributes` for querying the PTX version at runtime.

This will no longer work when we migrate to the alternative approach described here: https://github.com/NVIDIA/cub/issues/556

**Comment by zkhatami** [comment]
Is there any release targeted for #556? If its unknown, then might be beneficial from user perspective to have this fix with current CUB design. 
In this proposal, is there anyway to test for the GPU incompatibility in the very beginning, like the one that we currently have with using the failure on cudaFuncGetAttributes?

**Comment by jrhemstad** [comment]
> have this fix with current CUB design.

I agree, however, I'm not sure about our ability to preserve this functionality into the future. 

> In this proposal, is there anyway to test for the GPU incompatibility in the very beginning, like the one that we currently have with using the failure on cudaFuncGetAttributes?

That's the problem, I don't think we do.

File: cub/util_device.cuh [hunk 3]
```diff
@@ -348,6 +366,19 @@ CUB_RUNTIME_FUNCTION inline cudaError_t PtxVersionUncached(int& ptx_version)
...
         result = cudaFuncGetAttributes(&empty_kernel_attrs,
                                        reinterpret_cast<void*>(empty_kernel));
+        // Any failure means either no GPU is available or the provided device
+        // binary does not match the generated function code
+        if (result != cudaSuccess) {
+            int sm_version;
+            int device;
+            if (!CubDebug(cudaGetDevice(&device))) {
+                if (!CubDebug(SmVersionUncached(sm_version, device))) {
+                    printf("Incompatible GPU: you are trying to run this program on sm_%d%d, "
```
Comment (senior-zero) [inline cub/util_device.cuh:376]: Should we include something like stdio in this header if we are going to use printf unconditionally?

Comment (senior-zero) [reply cub/util_device.cuh:376]: Also, should it be `fprintf(stderr, ...)` instead?

File: cub/util_device.cuh [hunk 4]
```diff
@@ -348,6 +366,19 @@ CUB_RUNTIME_FUNCTION inline cudaError_t PtxVersionUncached(int& ptx_version)
...
+                            "different from the one that it was compiled for\n",
```
Comment (senior-zero) [inline cub/util_device.cuh:377]: Optional: we could report actual architectures we were compiled against. We are using the list in any case for our namespace generation:
```cpp
#define CUB_DETAIL_MAGIC_NS_BEGIN inline namespace CUB_DETAIL_MAGIC_NS_NAME(CUB_VERSION, NV_TARGET_SM_INTEGER_LIST) {
#else // not defined(_NVHPC_CUDA)
#define CUB_DETAIL_MAGIC_NS_BEGIN inline namespace CUB_DETAIL_MAGIC_NS_NAME(CUB_VERSION, __CUDA_ARCH_LIST__) {
```
We could probably concatenate appropriate lists and stringify the result as a follow up at some point.

File: cub/util_device.cuh [hunk 5]
```diff
@@ -348,6 +366,19 @@ CUB_RUNTIME_FUNCTION inline cudaError_t PtxVersionUncached(int& ptx_version)
...
+                            sm_version/100, (sm_version%100)/10);
+                }
+            }
+        }
```
Comment (senior-zero) [inline cub/util_device.cuh:381]: Some style guides: 
```suggestion
        if (result != cudaSuccess) 
        {
          int sm_version{};
          int device{-1};

          if (CubDebug(cudaGetDevice(&device)) == cudaSuccess) 
          {
            if (CubDebug(SmVersionUncached(sm_version, device)) == cudaSuccess) 
            {
              printf("Incompatible GPU: you are trying to run this program on sm_%d%d, "
                     "different from the one that it was compiled for\n",
                     sm_version / 100, (sm_version % 100) / 10);
            }
          }
        }
```

**Review by senior-zero — CHANGES_REQUESTED** [review]
Thank you for the contribution! Some minor questions below

**Comment by dkolsen-pgi** [comment]
I agree with Jake that writing error messages to stdout or stderr is not the best way to report this problem.  The use cases that we care about are calling Thrust.  NVC++ stdpar never calls CUB directly.  Thrust usually reports device-side problems by throwing a `thrust::system_error` exception.  A lack of GPU or a GPU with an incompatible architecture should also be reported with a `thrust::system_error` exception, with a a custom message in the exception that clearly explains the problem.

I wonder if `get_ptx_version` in `thrust/system/cuda/detail/core/util.h` would be a better place to put this logic.

**Comment by zkhatami** [comment]
@jrhemstad @dkolsen-pgi @senior-zero 
I've moved my changes to thrust instead based on David's suggestion. Here is my pull request:
https://github.com/NVIDIA/thrust/pull/1848

Still, I'm not able to add any reviewers here. The Reviewers button is still disabled for me.
Would you please take a look at the changes and share your comments with me? Thanks!

**Comment by senior-zero** [comment]
Hello @zkhatami!

> I've moved my changes to thrust instead based on David's suggestion. 

If the change is now on the thrust part, can this PR be closed? 

> Still, I'm not able to add any reviewers here. 
The thrust PR seems to have 3 reviewers at the moment. Let me know if I can help requesting reviews from other maintainers.

**Comment by zkhatami** [comment]
I'm closing this pull since I've moved the changes to thrust.

**Pull request closed without merging**